File: crypto/src/main/java/io/oversec/one/crypto/AppsReceiver.java
package io.oversec.one.crypto;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;

import java.util.HashSet;
import java.util.Set;

public class AppsReceiver extends BroadcastReceiver {
    @Override
    public void onReceive(Context context, Intent intent) {
        try {
            String packagename = intent.getData().getEncodedSchemeSpecificPart();
            fire(context, intent.getAction(), packagename);
        } catch (Exception ex) {
            ex.printStackTrace();
        }
    }

    private static final Set<IAppsReceiverListener> mListeners = new HashSet<>();

    public interface IAppsReceiverListener {
        void onAppChanged(Context ctx, String action, String packagename);
    }

    private static synchronized void fire(Context ctx, String action, String packagename) {
        for (IAppsReceiverListener listener : mListeners) {
            listener.onAppChanged(ctx, action, packagename);
        }
    }

    public static synchronized void addListener(IAppsReceiverListener listener) {
        mListeners.add(listener);
    }

    public static synchronized void removeListener(IAppsReceiverListener listener) {
        mListeners.remove(listener);
    }
}

File: crypto/src/main/java/io/oversec/one/crypto/AppsReceiver.kt
package io.oversec.one.crypto

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import java.util.HashSet

class AppsReceiver : BroadcastReceiver() {

    override fun onReceive(context: Context, intent: Intent) {
        try {
            intent?.data?.encodedSchemeSpecificPart?.let {
                packagename -> fire(context, intent.action, packagename )
            }
        } catch (ex: Exception) {
            ex.printStackTrace()
        }
    }

    interface IAppsReceiverListener {
        fun onAppChanged(ctx: Context, action: String, packagename: String)
    }

    companion object {

        private val mListeners = HashSet<IAppsReceiverListener>()

        @Synchronized
        private fun fire(ctx: Context, action: String, packagename: String) {
            mListeners.forEach {
                it.onAppChanged(ctx, action, packagename)
            }
        }

        @Synchronized
        fun addListener(listener: IAppsReceiverListener) {
            mListeners.add(listener)
        }

        @Synchronized
        fun removeListener(listener: IAppsReceiverListener) {
            mListeners.remove(listener)
        }
    }
}
